lib/src/html/html_generator.dart

```diff
@@ -66,22 +69,61 @@ class HtmlGenerator extends Generator {
 
   /// Actually write out the documentation for [package].
   /// Stores the HtmlGeneratorInstance so we can access it in [writtenFiles].
-  Future generate(Package package, String outputDirectoryPath) {
+  Future generate(Package package, String outputDirectoryPath) async {
     assert(_instance == null);
-    _instance = new HtmlGeneratorInstance(
-        _options, _templates, package, outputDirectoryPath, _onFileCreated);
-    return _instance.generate();
+
+    var enabled = true;
+    void write(String filePath, Object content, {bool allowOverwrite}) {
+      allowOverwrite ??= false;
+      if (!enabled) {
+        throw new StateError('`write` was called after `generate` completed.');
+      }
+      // If you see this assert, we're probably being called to build non-canonical
+      // docs somehow.  Check data.self.isCanonical and callers for bugs.
+      assert(allowOverwrite || !writtenFiles.contains(filePath));
+
+      var file = new File(p.join(outputDirectoryPath, filePath));
+      var parent = file.parent;
+      if (!parent.existsSync()) {
+        parent.createSync(recursive: true);
```

Comment:
RE ensuring the directory is created: it also happens here 😄

Reply:
Seems as though this is the only place it is really needed.